src/libosd/include/osd/cl_dem_uart.h

```diff
@@ -0,0 +1,116 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef OSD_CL_DEM_UART_H
+#define OSD_CL_DEM_UART_H
+
+#include <stdlib.h>
+
+#include <osd/osd.h>
+#include <osd/packet.h>
+#include <osd/hostmod.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+
+/**
+ * A osd_dem_uart_event represents one or more characters.
+ */
+struct osd_dem_uart_event {
+	uint8_t num_chars;
+	uint8_t chars[0];
+};
+
+typedef void (*osd_cl_dem_uart_handler_fn)(
+	void * /*cb_arg*/, struct osd_dem_uart_event * /*event to be handled*/);
+
+struct osd_dem_uart_event_handler {
+	osd_cl_dem_uart_handler_fn cb_fn;
+	void *cb_arg;
+};
+
+struct osd_dem_uart_desc {
+	unsigned int di_addr;
+};
+
+/**
+ * Obtain information about a UART Device-Emulation-Module (DEM)
+ *
+ * @param hostmod_ctx the host module handling the communication
+ * @param dem_uart_di_adr DI address of the UART-DEM to describe
+ * @param dem_uart_desc a pre-allocated dem_uart descriptor for the result
+ *
+ * @return OSD_OK on success, any other value indicates an error
+ */
+osd_result osd_cl_dem_uart_get_desc(struct osd_hostmod_ctx *hostmod_ctx,
+									uint16_t dem_uart_di_addr,
+									struct osd_dem_uart_desc *dem_uart_desc);
+
+/**
+ * Handle a packet received from a UART Device-Emulation-Module (DEM)
+ *
+ * This function should be passed to osd_hostmod_new. It takes care of
+ * converting the osd_packet into a osd_dem_uart_event and calling the
+ * associated handler function.
+ *
+ * @param handler the event handler that should handle this event
+ * @param pkg the osd_packet to be handled
+ *
+ * @return OSD_OK on success, any other value indicates an error
+ */
+osd_result osd_cl_dem_uart_receive_event(void *handler, struct osd_packet *pkg);
+
+/**
+ * Send the given event to the UART Device-Emulation-Module (DEM)
+ *
+ * For now an event only consists of a single char, later on that will change.
+ * The event is converted to an osd_packet with the dest and src fields
+ * set to dem_uart_di_addr and hostmod_di_addr respectively.
+ *
+ * @param dem_uart_ctx the context in which this event should be handled
+ * @param ev the new event that was raised on the host
+ *
+ * @return OSD_OK on success, any other value indicates an error
+ */
+osd_result osd_cl_dem_uart_send_event(struct osd_hostmod_ctx *hostmod_ctx,
+									  struct osd_dem_uart_event *ev,
+									  uint16_t dem_uart_di_addr);
+
+/**
+ * Create a new event big enough to hold at most num_chars chars.
+ *
+ * @param event pointer to a event-pointer to be filled with the new event
+ * @param num_chars the maximum number of chars this event should be able to hold
+ *
+ * @return OSD_OK on success, any other value indicates an error
+ */
+osd_result osd_cl_dem_uart_event_new(struct osd_dem_uart_event **event,
```

Comment:
I don't think you need this function once you remove it from the public API of `osd_cl_dem_uart_send_event()`.

Even less if you follow the suggestion to get rid of the `osd_dem_uart_event` entirely.